src/value_generators/include/value_generators/DocumentGenerator.hpp

```diff
@@ -57,6 +62,35 @@ struct GeneratorArgs {
     ActorId actorId;
 };
 
+class Appendable {
+public:
+    virtual ~Appendable() = default;
+    virtual void append(const std::string& key, bsoncxx::builder::basic::document& builder) = 0;
+    virtual void append(bsoncxx::builder::basic::array& builder) = 0;
+};
+
+using UniqueAppendable = std::unique_ptr<Appendable>;
+
+template <class T>
+class Generator : public Appendable {
```

Comment:
do you actually need the def here or just to forward-declare it? I think you can get away with just the decl since you operate by ptr.

Reply:
Yeah, might be good to sit with you on this one. At the end of the day I just need to be able to forward declare
UniqueGenerator<double> doubleGenerator(const Node& node, GeneratorArgs generatorArgs);
What is the minimum thing for that? 
UniqueGenerator is introduced by a using. 
Could I do

template <class t> class Generator;
using UniqueGenerator = std::unique_ptr<Generator<T>>;
UniqueGenerator<int64_t> intGenerator(const Node& node, GeneratorArgs generatorArgs);
UniqueGenerator<double> doubleGenerator(const Node& node, GeneratorArgs generatorArgs);

Or maybe 
class Appendable;
template <class t> class Generator : public Appendable;
using UniqueGenerator = std::unique_ptr<Generator<T>>;
UniqueGenerator<int64_t> intGenerator(const Node& node, GeneratorArgs generatorArgs);
UniqueGenerator<double> doubleGenerator(const Node& node, GeneratorArgs generatorArgs);

I supposed we could also pimpl it for exposing the intGenerator. That might be cleanest. What do you think?